### DISCUSSION TITLE
How to create windows executable of ifctester?

### DISCUSSION THREAD

**Discussion by Ghevi** [IfcOpenShell/IfcOpenShell#5844] [Q&A]
I would like to distribute the `ifctester` as an executable and i've tried using pyinstaller.
I've downloaded the repository as a zip and then opened a terminal in `src/ifctester`.

Running the script with `python -m ifctester example.ids example.ifc` works. (those are example files i've put in the folder)

But then I tried to create the executable with this command: `pyinstaller ifctester/__main__.py`. 
When running the exe i get the error:
`File "ifctester\__main__.py", line 24, in <module>
ImportError: attempted relative import with no known parent package`
That line is `from . import ids`, so probably is not correct to bundle `__main__.py` with pyinstaller?

I've tried `from ifctester.ids import ids` but again it cannot find the import, giving this error:
`ImportError: cannot import name 'ids' from 'ifctester.ids' (C:\Users\ghevi.sartor\Desktop\IfcOpenShell\IfcOpenshell\src\`

**Comment by aothms** [comment]
This is likely a generic pyinstaller issue that other packages also encounter.

> attempted relative import with no known parent package

When you invoke `python -m ifctester` you invoke it as a module (-m). But when you treat `ifctester/__main__.py` as a standalone script python gets confused because now it no longer behaves as a module. 

After 15y of python experience, modules still trip me up, but this sounds like something you could find generic advise for on the internet, on how to package a module in pyinstaller.

**Reply by Ghevi** [reply]
Hi, after some more tests, reading bits online about pyinstaller i've managed to do it, i will post how i did it soon

**Comment by Ghevi** [comment]
It has something to do with pyinstaller and relative imports and also not including some files in the correct place.
What I did was:

- Run `pip install xmlschema`
- Run `pip install pystache` 
- Added in `__main__.py` file, right below the imports, `def main():` and indented everything below
- Created a file that I've called `cli.py` in the root ifctester folder with these lines: `from ifctester.__main__ import main 
main()`
- In the root folder run this command in cmd: `pyinstaller --onefile --add-data "ifctester/ids.xsd;ifctester" --add-data "ifctester/templates;ifctester/templates" cli.py`
- In the dist folder run the exe with `cli.exe fail.ids fail.ifc -r Html -o report.html` (i've put those files as an example)

Unfortunately this causes the `__main__.py` to do nothing with the original `python -m ifctester` command, since eveything is inside a function with these changes.
The `--add-data` paths was necessary because at runtime those are the paths (the ones after ';') where the exe tried to search for those files

**Reply by aothms** [reply]
Thanks for coming back to post a solution that works 👍 

> Unfortunately this causes the __main__.py to do nothing with the original python -m ifctester command, since eveything is inside a function with these changes.

I don't understand the implications of that or what you mean precisely
